File: golang/tests/pkg/integration/ws_register_callback_integration_test.go
package integration

import (
	"os"
	"testing"
	"time"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/browser"
	"github.com/playwright-community/playwright-go"
)

func TestWsRegisterCallback_ShouldReceiveCaptchaPush(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("AGENTBAY_API_KEY environment variable not set")
	}

	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("new agentbay: %v", err)
	}

	created, err := client.Create(agentbay.NewCreateSessionParams())
	if err != nil {
		t.Fatalf("create session: %v", err)
	}
	if created == nil || created.Session == nil {
		t.Fatalf("create session returned nil session")
	}
	session := created.Session
	defer func() { _, _ = session.Delete(false) }()

	wsClientAny, err := session.GetWsClient()
	if err != nil {
		t.Fatalf("get ws client: %v", err)
	}
	type wsPushClient interface {
		RegisterCallback(target string, callback func(payload map[string]interface{})) func()
		Close() error
	}
	wsClient, ok := wsClientAny.(wsPushClient)
	if !ok {
		t.Fatalf("unexpected ws client type: %T", wsClientAny)
	}
	defer func() { _ = wsClient.Close() }()

	gotCh := make(chan map[string]interface{}, 1)
	wsClient.RegisterCallback("wuying_cdp_mcp_server", func(payload map[string]interface{}) {
		select {
		case gotCh <- payload:
		default:
		}
	})

	opt := browser.NewBrowserOption()
	opt.UseStealth = true
	opt.SolveCaptchas = true
	initialized, err := session.Browser.Initialize(opt)
	if err != nil {
		t.Fatalf("init browser: %v", err)
	}
	if !initialized {
		t.Fatalf("init browser returned false")
	}
	endpointURL, err := session.Browser.GetEndpointURL()
	if err != nil {
		t.Fatalf("get endpoint url: %v", err)
	}

	pw, err := playwright.Run()
	if err != nil {
		t.Fatalf("playwright run: %v", err)
	}
	defer func() { _ = pw.Stop() }()

	remote, err := pw.Chromium.ConnectOverCDP(endpointURL)
	if err != nil {
		t.Fatalf("connect over cdp: %v", err)
	}
	defer func() { _ = remote.Close() }()

	ctxs := remote.Contexts()
	if len(ctxs) == 0 {
		t.Fatalf("no browser contexts")
	}
	page, err := ctxs[0].NewPage()
	if err != nil {
		t.Fatalf("new page: %v", err)
	}
	defer func() { _ = page.Close() }()

	_, err = page.Goto("https://passport.ly.com/Passport/GetPassword", playwright.PageGotoOptions{
		WaitUntil: playwright.WaitUntilStateDomcontentloaded,
	})
	if err != nil {
		t.Fatalf("goto: %v", err)
	}
	input, err := page.WaitForSelector("#name_in", playwright.PageWaitForSelectorOptions{Timeout: playwright.Float(10000)})
	if err != nil {
		t.Fatalf("wait selector: %v", err)
	}
	_ = input.Click()
	_ = input.Fill("")
	_ = input.Type("13000000000")
	page.WaitForTimeout(1000)
	_ = page.Click("#next_step1")

	select {
	case payload := <-gotCh:
		if payload == nil {
			t.Fatalf("nil payload")
		}
		if payload["target"] != "wuying_cdp_mcp_server" {
			t.Fatalf("unexpected payload target: %v", payload["target"])
		}
		data, _ := payload["data"].(map[string]interface{})
		if data == nil {
			t.Fatalf("payload.data is not object: %v", payload["data"])
		}
		codeAny := data["code"]
		code, _ := codeAny.(float64)
		if code != 201 && code != 202 {
			t.Fatalf("unexpected data.code: %v", codeAny)
		}
	case <-time.After(180 * time.Second):
		t.Fatalf("timeout waiting for ws push")
	}
}

File: java/agentbay/src/integration-test/java/com/aliyun/agentbay/test/WsRegisterCallbackIntegrationTest.java
package com.aliyun.agentbay.test;

import com.aliyun.agentbay.AgentBay;
import com.aliyun.agentbay._internal.WsClient;
import com.aliyun.agentbay.browser.BrowserOption;
import com.aliyun.agentbay.model.SessionResult;
import com.aliyun.agentbay.session.CreateSessionParams;
import com.aliyun.agentbay.session.Session;
import com.microsoft.playwright.Browser;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.Playwright;
import org.junit.AfterClass;
import org.junit.Assume;
import org.junit.BeforeClass;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import static org.junit.Assert.*;

public class WsRegisterCallbackIntegrationTest {
    private static AgentBay agentBay;
    private static Session session;

    @BeforeClass
    public static void setUp() throws Exception {
        String apiKey = System.getenv("AGENTBAY_API_KEY");
        Assume.assumeTrue(apiKey != null && !apiKey.isEmpty());

        agentBay = new AgentBay(apiKey);

        CreateSessionParams params = new CreateSessionParams();
        SessionResult created = agentBay.create(params);
        assertTrue("Failed to create session: " + created.getErrorMessage(), created.isSuccess());
        session = created.getSession();
        assertNotNull(session);
        assertNotNull("wsUrl should not be empty", session.getWsUrl());
        assertTrue("wsUrl should not be empty", !session.getWsUrl().isEmpty());
    }

    @AfterClass
    public static void tearDown() {
        if (session != null && agentBay != null) {
            try {
                agentBay.delete(session, false);
            } catch (Exception e) {
                System.err.println("Cleanup error: " + e.getMessage());
            }
        }
    }

    @Test
    public void testWsRegisterCallbackShouldReceiveCaptchaPushE2E() throws Exception {
        WsClient wsClient = session.getWsClient();
        assertNotNull(wsClient);

        CountDownLatch latch = new CountDownLatch(1);
        List<Map<String, Object>> received = new ArrayList<>();
        wsClient.registerCallback("wuying_cdp_mcp_server", payload -> {
            received.add(payload);
            latch.countDown();
        });

        BrowserOption option = new BrowserOption();
        option.setUseStealth(true);
        option.setSolveCaptchas(true);
        boolean ok = session.getBrowser().initialize(option);
        assertTrue("Failed to initialize browser", ok);

        String endpointUrl = session.getBrowser().getEndpointUrl();
        assertNotNull(endpointUrl);
        assertTrue(endpointUrl.startsWith("ws://") || endpointUrl.startsWith("wss://"));

        try (Playwright playwright = Playwright.create()) {
            Browser browser = playwright.chromium().connectOverCDP(endpointUrl);
            Page page = browser.contexts().get(0).newPage();
            page.navigate("https://passport.ly.com/Passport/GetPassword");
            page.waitForSelector("#name_in").fill("13000000000");
            page.waitForTimeout(1000);
            page.click("#next_step1");

            boolean got = latch.await(180, TimeUnit.SECONDS);
            assertTrue("Timeout waiting for WS push callback", got);
        }

        assertFalse("Expected at least one push message", received.isEmpty());
        Map<String, Object> first = received.get(0);
        assertEquals("wuying_cdp_mcp_server", first.get("target"));
        assertNotNull(first.get("requestId"));
        Object dataObj = first.get("data");
        assertTrue("payload.data should be an object", dataObj instanceof Map);
        Map<?, ?> data = (Map<?, ?>) dataObj;
        Object code = data.get("code");
        assertTrue("unexpected data.code: " + code, code != null && (Integer.valueOf(201).equals(code) || Integer.valueOf(202).equals(code)));
    }
}

